app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,107 @@
+module MiqServer::WorkerManagement::Monitor::Kubernetes
+  extend ActiveSupport::Concern
+  attr_accessor :pod_resource_version
+  attr_reader :current_pods
+
+  def initialize(*args)
+    @current_pods = Concurrent::Hash.new
+    super
+  end
+
+  def cleanup_failed_deployments
+    ensure_pod_monitor_started
+    delete_failed_deployments
+  end
+
+  def failed_deployments(restart_count = 5)
+    # TODO: This logic might flag deployments that are hitting memory/cpu limits or otherwise not really 'failed'
+    current_pods.values.select { |h| h[:last_state_terminated] && h.fetch(:container_restarts, 0) > restart_count }.collect { |h| h[:label_name] }
+  end
+
+  private
+
+  def start_pod_monitor
+    @monitor_thread ||= Thread.new { monitor_pods }
+  end
+
+  def ensure_pod_monitor_started
+    if @monitor_thread.nil? || !@monitor_thread.alive?
+      if !@monitor_thread.nil? && @monitor_thread.status.nil?
+        dead_thread, @monitor_thread = @monitor_thread, nil
+        _log.info("Waiting for the Monitor Thread to exit...")
+        dead_thread.join
+      end
+
+      start_pod_monitor
+    end
+  end
+
+  def delete_failed_deployments
+    # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+    # See ContainerOrchestrator#get_pods
+    failed_deployments.each do |failed|
+      orchestrator.delete_deployment(failed)
+    end
+  end
+
+  def orchestrator
+    @orchestrator ||= ContainerOrchestrator.new
+  end
+
+  def monitor_pods
+    loop do
+      current_pods.clear
+      collect_initial_pods
+
+      # watch_for_pod_events doesn't return unless an error caused us to break out of it, so we'll reset and start over again
+      watch_for_pod_events
+    end
+  end
+
+  def collect_initial_pods
+    pods = orchestrator.get_pods
+    pods.each { |p| save_pod(p) }
+    self.pod_resource_version = pods.resourceVersion
+  end
+
+  def watch_for_pod_events
+    orchestrator.watch_pods(pod_resource_version).each do |event|
+      case event.type.downcase
+      when "added", "modified"
+        save_pod(event.object)
+      when "deleted"
+        delete_pod(event.object)
+      when "error"
+        if status = event.object
+          # ocp 3 appears to return 'ERROR' watch events with the object containing the 410 code and "Gone" reason like below:
+          # #<Kubeclient::Resource type="ERROR", object={:kind=>"Status", :apiVersion=>"v1", :metadata=>{}, :status=>"Failure", :message=>"too old resource version: 199900 (27177196)", :reason=>"Gone", :code=>410}>
+          log_pod_error_event(status.code, status.message, status.reason)
+        end
+
+        self.pod_resource_version = nil
```

Comment:
Is there much benefit to having `:pod_resource_version` be an instance var when it isn't used outside of [this loop](https://github.com/ManageIQ/manageiq/pull/20444/files#diff-84282d4bc36b3c58630473ba1b4099c4R52-R58)?

Looks like you could return it from `collect_initial_pods` and pass it in to `watch_for_pod_events` then you don't have to worry about setting it to `nil` (unless I'm missing another context where it is being used)